samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartImpl.java

```diff
@@ -16,94 +16,109 @@
 import com.akkaserverless.javasdk.eventsourcedentity.CommandContext;
 import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntity;
 import com.example.shoppingcart.ShoppingCartApi;
 import com.google.protobuf.Empty;
 
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 @EventSourcedEntity(entityType = "eventsourced-shopping-cart")
-public class ShoppingCartImpl extends ShoppingCartInterface {
+public class ShoppingCartImpl extends ShoppingCartInterface2 {
   @SuppressWarnings("unused")
   private final String entityId;
 
-  private final Map<String, ShoppingCartApi.LineItem> cart = new LinkedHashMap<>();
-
   public ShoppingCartImpl(@EntityId String entityId) {
     this.entityId = entityId;
   }
 
   @Override
-  public ShoppingCartDomain.Cart snapshot() {
-    return ShoppingCartDomain.Cart.newBuilder()
-        .addAllItems(cart.values().stream().map(this::convert).collect(Collectors.toList()))
-        .build();
+  protected ShoppingCartDomain.Cart emptyState() {
+    return ShoppingCartDomain.Cart.getDefaultInstance();
   }
 
   @Override
-  public void handleSnapshot(ShoppingCartDomain.Cart cart) {
-    this.cart.clear();
-    for (ShoppingCartDomain.LineItem item : cart.getItemsList()) {
-      this.cart.put(item.getProductId(), convert(item));
+  public Effect<Empty> addItem(
+          ShoppingCartDomain.Cart currentState,
+          ShoppingCartApi.AddLineItem command,
+          CommandContext<Empty, ShoppingCartDomain.Cart> context) {
+    if (command.getQuantity() <= 0) {
+      return context.effectBuilder().failure("Cannot add negative quantity of to item" + command.getProductId());
     }
-  }
 
-  @Override
-  protected Empty addItem(ShoppingCartApi.AddLineItem item, CommandContext ctx) {
-    if (item.getQuantity() <= 0) {
-      throw ctx.fail("Cannot add negative quantity of to item" + item.getProductId());
-    }
-    ctx.emit(
-        ShoppingCartDomain.ItemAdded.newBuilder()
+    ShoppingCartDomain.ItemAdded event = ShoppingCartDomain.ItemAdded.newBuilder()
             .setItem(
-                ShoppingCartDomain.LineItem.newBuilder()
-                    .setProductId(item.getProductId())
-                    .setName(item.getName())
-                    .setQuantity(item.getQuantity())
-                    .build())
-            .build());
-    return Empty.getDefaultInstance();
+                    ShoppingCartDomain.LineItem.newBuilder()
+                            .setProductId(command.getProductId())
+                            .setName(command.getName())
+                            .setQuantity(command.getQuantity())
+                            .build())
+            .build();
+
+    return context.effectBuilder()
+            .emitEvent(event)
+            .thenReply(newState -> Empty.getDefaultInstance());
```

Comment:
This is looking much better. 

I wonder what we could do to remove the context at all. 

For instance, in Akka Typed, the entity needs to implement a given type. Could we do the same here? 

```java
abstract class ShoppingCartInterface2 implements EventSourcedEntity<Cart>
class ShoppingCartImpl extends ShoppingCartInterface2
```

`EventSourcedEntity` would have the effect builders already typed on `Cart` and will also have the `commandHandler` and `eventHandler` entry points to be implemented.

`ShoppingCartInterface2` would implement `commandHandler` and `eventHandler` and do the dispatch to the methods the user needs to implement. And we make handler final in the interface. The user only touch the method to dispatch to.

For the return type, I don't think the builder needs to have a type argument, because the compiler will force the user to finish it with the right type, ie: `.thenReply(newState -> "I'm done");` won't compile.

We are left with the data that is provide by the context, command name, command id and seqNum (for event sourcing). 

Would that be a good idea to wrap the command instead? 

```java
class ShoppingCartImpl extends implements EventSourcedEntity<Cart> {
  
  public Effect<Empty> addItem(
    ShoppingCartDomain.Cart currentState,
    Command<ShoppingCartApi.AddLineItem> input) {
	  
    // use wrap to access command and other call related info
    input.command() // returns ShoppingCartApi.AddLineItem
    input.commandName()
    input.commandId()
	    
    return effectBuilder() // already typed on Cart because of interface
             .emitEvent(event)
             // compile enforces return type
             .thenReply(newState -> Empty.getDefaultInstance()); 

	  }
}
```

Reply:
I think we need the `implements EventSourcedEntity<Cart>` anyway, for other reasons. I think I mentioned that in relation to the `emptyState`.

I like the idea of having the `effectBuilder()` (or `effects()` as was suggested elsewhere) in that base interface. I have to try if that really works with the reply type as you say.

I'm skeptical to the `Command` wrapping. One less parameter, but it becomes more indirection for the most common usage, which is accessing the command message. The context parameter can often be ignored.

Reply:
Having the interface typed looks good, and for empty / initial state, and with the effects builder in the base rather than context.

Reply:
> I'm skeptical to the Command wrapping. One less parameter, but it becomes more indirection for the most common usage, which is accessing the command message. The context parameter can often be ignored.

Yeah, you may be right. I just dislike having params that are not used. Specially if they are often ignored. They are always available and often ignored. Just noise. 

I wish we didn't have that at all. Do we really need a `commandName` if we have already the command? 

And what is a `commandId` exactly? 
If that's something generated by the framework, why do I need such a thing in my domain? And if it's something that I need in my domain than that should be a field in my command.

Reply:
I don't think we will be able to fully get rid of the context. There are several metadata related things that should be accessible, such as
```
CommandContext.commandName
CommandContext.commandId
MetadataContext.metadata

EntityContext.entityId
Context.serviceCallFactory

EventContext.sequenceNumber


ActionContext.eventSubject

UpdateHandlerContext.eventSubject
UpdateHandlerContext.commandName
ViewContext.viewId
```

Let me try the reply type first...

Reply:
Ok, let' go by pieces. 

I believe that some of then can be a constructor parameter or deemed obsolete (or not crucial). We will see..

Reply:
Moved to the base class approach: 23c7fb9
Looking good.

I'll give the context some more thought. We could always give access to it via the base class instead of a parameter.

Reply:
I removed the context parameter, and it's accessed via the base class instead. Some runtime caveats but that is worth it. 05239d2